proxy/src/main/java/com/wavefront/agent/preprocessor/SpanSanitizeTransformer.java

```diff
@@ -0,0 +1,77 @@
+package com.wavefront.agent.preprocessor;
+
+import com.google.common.base.Function;
+import com.wavefront.common.Pair;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import wavefront.report.Annotation;
+import wavefront.report.Span;
+
+/**
+ * Sanitize spans (e.g., span source and tag keys) according to the same rules that are applied at
+ * the SDK-level.
+ *
+ * @author Han Zhang (zhanghan@vmware.com)
+ */
+public class SpanSanitizeTransformer implements Function<Span, Span> {
+  private final PreprocessorRuleMetrics ruleMetrics;
+
+  public SpanSanitizeTransformer(final PreprocessorRuleMetrics ruleMetrics) {
+    this.ruleMetrics = ruleMetrics;
+  }
+
+  @Nullable
+  @Override
+  public Span apply(@Nullable Span span) {
+    long startNanos = ruleMetrics.ruleStart();
+    boolean ruleApplied = false;
+    Pair<String, Boolean> sanitizedSource = sanitize(span.getSource());
+    if (sanitizedSource._2) {
+      span.setSource(sanitizedSource._1);
+      ruleApplied = true;
+    }
+    if (span.getAnnotations() != null) {
+      for (Annotation a : span.getAnnotations()) {
+        Pair<String, Boolean> sanitizedKey = sanitize(a.getKey());
+        if (sanitizedKey._2) {
+          a.setKey(sanitizedKey._1);
+          ruleApplied = true;
+        }
+      }
+    }
+    if (ruleApplied) {
+      ruleMetrics.incrementRuleAppliedCounter();
+    }
+    ruleMetrics.ruleEnd(startNanos);
+    return span;
+  }
+
+  /**
+   * Sanitize a string so that every invalid character is replaced with a dash.
+   *
+   * @param s The string to sanitize.
+   * @return A {@link Pair} containing the resulting string and a boolean indicating whether any
+   * sanitization was applied.
+   */
+  private Pair<String, Boolean> sanitize(String s) {
```

Comment:
if this is defined in tier 1 SDK, should we just use that instead of defining it again?

Reply:
the method in the SDK wraps the string in quotes, but i think doing that here would be incorrect.